lib/link_accounts_page.dart

```diff
@@ -1,134 +1,194 @@
+        _uiController.add(UIStatus.done);
+        return;
+      }
 
-    await user.linkWithCredential(credential);
+      final googleAuth = await _googleUser.authentication;
 
-    // we are linked so reset the UI
-    yield 0;
-  } catch (error) {
-    // reset the UI and display an alert
+      final credential = GoogleAuthProvider.getCredential(
+        accessToken: googleAuth.accessToken,
+        idToken: googleAuth.idToken,
+      );
 
-    yield 0;
-    // errors with code kSignInCanceledError are swallowed by the
-    // GoogleSignIn.signIn() method so we can assume anything caught here
-    // is unexpected and for display
-    _showDialog(context, error.toString());
+      await user.linkWithCredential(credential);
+
+      // we are linked so reload the user's data and reset the UI
+      await user.reload();
+      _uiController.add(UIStatus.done);
+    } catch (error) {
+      // reset the UI and display an alert
+      _uiController.add(UIStatus.error);
+      // errors with code kSignInCanceledError are swallowed by the
+      // GoogleSignIn.signIn() method so we can assume anything caught here
+      // is unexpected and for display
+      rethrow;
+    }
   }
-}
 
-Stream<int> _linkFacebook(BuildContext context, FirebaseUser user) async* {
-  try {
-    final facebookLogin = FacebookLogin();
-    final result = await facebookLogin.logIn(['email']);
+  Future<void> _linkFacebook(FirebaseUser user) async {
+    // signal to change UI
+    _uiController.add(UIStatus.loading);
+
+    try {
+      final facebookLogin = FacebookLogin();
+      final result = await facebookLogin.logIn(['email']);
+
+      switch (result.status) {
+        case FacebookLoginStatus.loggedIn:
+
+          /// the auth info will be picked up by the listener on [onAuthStateChanged]
+          /// and emitted by [streamOfStateChanges]
+
+          final credential = FacebookAuthProvider.getCredential(
+              accessToken: result.accessToken.token);
+
+          await user.linkWithCredential(credential);
+
+          // we are signed in so reload the user's data and reset the UI
+          await user.reload();
+          _uiController.add(UIStatus.done);
+          break;
+        case FacebookLoginStatus.cancelledByUser:
+          _uiController.add(UIStatus.done);
+          break;
+        case FacebookLoginStatus.error:
+          _uiController.add(UIStatus.error);
+          throw result.errorMessage;
+          break;
+      }
+    } catch (error) {
+      // reset the UI and display an alert
+      _uiController.add(UIStatus.error);
+      rethrow;
+    }
+  }
 
-    switch (result.status) {
-      case FacebookLoginStatus.loggedIn:
+  Widget _buildLoadingUI() {
+    return CircularProgressIndicator();
+  }
 
-        /// the auth info will be picked up by the listener on [onAuthStateChanged]
-        /// and emitted by [streamOfStateChanges]
+  Widget _buildLoadedUI(BuildContext context, FirebaseUser user) {
```

Comment:
Is the choice of functions that return widgets over subclassing your preference? Are you aware of the reasons to prefer subclassing?  I think it's ultimately personal preference but I think it's probably worth having a discussion and bouncing opinions around till we're all happy.  Here's a reference with some of the reasons why some prefer subclassing: https://github.com/flutter/flutter/issues/19269

Reply:
I don't have any particular preference for functional widgets versus subclassing, I guess I just wrote them that way because I'm used to functional components in React. I've read through the issue you posted and I think that subclassing is the correct thing to do in this case. I've made the necessary modifications in the latest commits.